[fix] Detect missing nodes in subgraphs

viva-jinyi opened the pull request:
## Summary
This PR fixes an issue where missing nodes inside subgraphs were not being detected and reported to users. The missing node warning dialog would only appear for missing nodes at the top level of the workflow, but not for nodes within subgraphs.

## Changes
- Modified `loadGraphData` in `app.ts` to recursively check nodes in subgraphs for missing node types
- Added context to missing node error messages showing the subgraph location (e.g., "MISSING_NODE (in subgraph 'MySubgraph')")
- Created test workflow `missing_nodes_in_subgraph.json` with a missing node inside a subgraph
- Added Playwright test to verify the warning appears for missing nodes in subgraphs

## Testing
- Created new Playwright test that verifies missing nodes in subgraphs are detected
- All existing tests pass
- Manual testing confirms the warning dialog now shows missing nodes from subgraphs with proper context

## Related Issues
Fixes #4618

## Screenshots
[screen-capture.webm](https://github.com/user-attachments/assets/494ae0d4-edd9-4f36-b2c0-03612a6490ba)



┆Issue is synchronized with this [Notion page](https://www.notion.so/PR-4639-fix-Detect-missing-nodes-in-subgraphs-2426d73d36508160ac5ce8cb558fcb01) by [Unito](https://www.unito.io)

viva-jinyi committed:
[fix] Detect missing nodes in subgraphs

- Modified loadGraphData to recursively check nodes in subgraphs
- Added context to missing node error messages showing subgraph location
- Created test workflow and Playwright test for missing nodes in subgraphs
- Fixed unrelated ESLint error in dialog.spec.ts

Fixes #4618

christian-byrne commented on src/scripts/app.ts:
```diff
@@ -1092,23 +1092,42 @@ export class ComfyApp {
-        embeddedModels.push(...selectedModels)
+    // Process nodes in subgraphs
+    if (graphData.definitions?.subgraphs) {
+      for (const subgraph of graphData.definitions.subgraphs as any[]) {
+        if (subgraph.nodes) {
```
it seems like `nodes` is not defined on `graphData.definitions.subgraphs` in the schema https://github.com/Comfy-Org/ComfyUI_frontend/blob/main/src/schemas/comfyWorkflowSchema.ts

But in the actual serialized workflows, the `nodes` property exists. I think this might be a mistake in the schema.

christian-byrne replied:
Looking into it now

christian-byrne replied:
I found the issue: https://github.com/Comfy-Org/ComfyUI_frontend/issues/4650.

viva-jinyi replied:
Removed the as any[] cast - replaced with a defensive type guard 'nodes' in subgraph &&  Array.isArray(subgraph.nodes) to ensure type safety

christian-byrne replied:
I am adding a typeguard to use so we don't need to cast to `any`: https://github.com/Comfy-Org/ComfyUI_frontend/pull/4651

viva-jinyi commented on src/scripts/app.ts:
```diff
@@ -1092,23 +1092,42 @@ export class ComfyApp {
+        // Find missing node types
+        if (!(n.type in LiteGraph.registered_node_types)) {
+          // Include context about subgraph location if applicable
+          const nodeTypeWithContext = path
+            ? `${n.type} (in subgraph '${path}')`
+            : n.type
+          missingNodeTypes.push(nodeTypeWithContext)
+          n.type = sanitizeNodeName(n.type)
```
Thank you for the review! I've updated the code to use the proper MissingNodeType format as you suggested.

viva-jinyi committed:
[fix] Address PR review comments for missing nodes in subgraphs

- Use proper MissingNodeType object format with hint field instead of
  modifying the node type string directly
- Remove 'as any[]' cast and use type guard for type safety when
  accessing subgraph nodes property
- Preserve original node type information while providing subgraph
  context through the standardized hint field

This maintains compatibility with the extension lifecycle hook API
and improves type safety by avoiding unsafe type assertions.

christian-byrne commented on src/scripts/app.ts:
```diff
@@ -1092,23 +1092,47 @@ export class ComfyApp {
+    processNodesRecursively(graphData.nodes)
+
+    // Process nodes in subgraphs
+    if (graphData.definitions?.subgraphs) {
```
Can you use this typeguard so that we don't need to use `any` in line 1096?

https://github.com/Comfy-Org/ComfyUI_frontend/blob/61c9341450f56f892cc7673b33de8a1fb799df3e/src/schemas/comfyWorkflowSchema.ts#L463-L478

This will cast to the correct type. It is necessary because of zod `lazy` on circular subgraph schema.

christian-byrne committed:
use type guard

christian-byrne reviewed (approved):
LGTM thanks for adding browser test
